### ISSUE TITLE
Decorator for generic class causes it to lose typing information

### ISSUE THREAD

**Issue by wch** [microsoft/pyright#3668]
I'm using a decorator that modifies the `__name__` and `__docstring__` of functions and callable class instances. It used to come up clean with pyright 1.1.256, but results in errors in 1.1.257.

Easiest to demonstrate with an example. This passes, as expected:

```py
from typing import Generic, TypeVar

T = TypeVar("T")

class Value(Generic[T]):
    def __init__(self, value: T) -> None:
        self._value: T = value

    def __call__(self) -> T:
        return self._value

v = Value(1)
```

*****

The code below is similar to my code. It passed in 1.1.256, but errors in 1.1.257:

```py
from typing import Any, Callable, Generic, TypeVar

T = TypeVar("T")
FuncType = Callable[..., Any]
FT = TypeVar("FT", bound=FuncType)

def decorate() -> Callable[[FT], FT]:
    def dec(fn: FT) -> FT:
        fn.__name__ = "new name"
        return fn

    return dec

@decorate()
class ValueDecorated(Generic[T]):
    def __init__(self, value: T) -> None:
        self._value: T = value

    def __call__(self) -> T:
        return self._value

vd = ValueDecorated(1)
```

The errors: 
```
  /.../test2.py:18:31 - error: TypeVar "T" appears only once in generic function signature (reportInvalidTypeVarUse)
  /.../test2.py:22:16 - error: Expression of type "T@__init__" cannot be assigned to return type "T@ValueDecorated"
    Type "T@__init__" cannot be assigned to type "T@ValueDecorated" (reportGeneralTypeIssues)
  /.../test2.py:25:1 - error: Type of "vd" is partially unknown
    Type of "vd" is "ValueDecorated[Unknown]" (reportUnknownVariableType)
3 errors, 0 warnings, 0 informations 
Completed in 0.45sec
```

Interestingly, it looks clean in VS Code with pyright 1.1.257 -- no underlining (though it's possible my editor isn't configured correctly).
<img width="456" alt="image" src="https://user-images.githubusercontent.com/86978/177646952-6866466c-e707-4282-8a6d-c80900efdaac.png">

*****

Now, instead of `TypeVar("FT", bound=FuncType)`, I could just use `TypeVar("F")`, but then it results in (expected) errors because it can't be sure that a `F` object has a `__name__`.

```py
from typing import Callable, Generic, TypeVar

F = TypeVar("F")
T = TypeVar("T")


def decorate() -> Callable[[F], F]:
    def dec(fn: F) -> F:
        fn.__name__ = "new name"
        return fn

    return dec


@decorate()
class ValueDecorated(Generic[T]):
    def __init__(self, value: T) -> None:
        self._value: T = value

    def __call__(self) -> T:
        return self._value


vd = ValueDecorated(1)
```

The error, which is expected:

```
  /.../test3.py:9:12 - error: Cannot assign member "__name__" for type "object*"
    Member "__name__" is unknown (reportGeneralTypeIssues)
```

*****

I'm not sure if I'm using the wrong type specification, or if there's a new bug in pyright. Any help would be appreciated.

**Comment by erictraut** [comment]
Thanks for the bug report. This will be fixed in the next release.

The problem is that the type variable `T` used in `__init__` was not being attributed to the class scope in this case. It was caused by a subtle interaction between various parts of the type evaluator logic when a class decorator was present. I've added a test case based on your code sample to prevent similar regressions in the future.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.258, which I just published. It will also be included in the next release of pylance.
